File: backend/services/scan_service.py
"""
YUMMY Backend - Scan Service
Background task để index codebase từ GitHub vào knowledge base.
"""

import os
import time
from config import DB, ALLOWED_EXTENSIONS
from services.ai_service import call_ai
from services.github_service import get_repo_info, get_repo_tree, github_raw


async def run_scan():
    """
    Background task: quét GitHub repo, tạo AI insights và project wiki.
    
    Flow:
        1. Lấy repo metadata (default branch)
        2. Lấy file tree, filter extensions + node_modules
        3. Đọc từng file, gộp thành chunks ~35KB
        4. Mỗi chunk → INDEXER agent tóm tắt
        5. Tất cả insights → ARCHITECT agent viết Project Wiki
    
    Poll status qua GET /kb/scan/status.
    """
    DB["scan_status"] = {"running": True, "text": "Kết nối GitHub API...", "progress": 0}
    DB["knowledge_base"] = {"tree": [], "insights": [], "project_summary": ""}

    try:
        ri = DB["repo_info"]
        max_limit = DB.get("max_scan_limit", 10000)

        # --- Step 1: Lấy repo metadata ---
        DB["scan_status"]["text"] = "Đọc thông tin repo..."
        repo_data = await get_repo_info(ri["owner"], ri["repo"])
        branch = repo_data["default_branch"]
        DB["repo_info"]["branch"] = branch  # cache branch để dùng sau

        # --- Step 2: Lấy file tree ---
        DB["scan_status"]["text"] = "Lấy danh sách files..."
        all_files = await get_repo_tree(ri["owner"], ri["repo"], branch)

        valid_files = [
            f for f in all_files
            if f["type"] == "blob"
            and "node_modules" not in f["path"]
            and ".git" not in f["path"]
            and os.path.splitext(f["path"])[1].lower() in ALLOWED_EXTENSIONS
        ][:max_limit]

        if not valid_files:
            DB["scan_status"] = {
                "running": False,
                "text": "⚠️ Không tìm thấy file nào phù hợp trong repo.",
                "progress": 0,
                "error": True
            }
            return

        # Khởi tạo tree với status "pending"
        DB["knowledge_base"]["tree"] = [
            {
                "path": f["path"],
                "name": f["path"].split("/")[-1],
                "status": "pending"
            }
            for f in valid_files
        ]

        # --- Step 3 & 4: Đọc file theo chunks, AI tóm tắt ---
        current_chunk = ""
        files_in_chunk = []
        insights = []
        total = len(valid_files)

        for i, file in enumerate(valid_files):
            progress = round((i / total) * 80)
            DB["scan_status"] = {
                "running": True,
                "text": f"Indexing [{file['path']}] ({i + 1}/{total})",
                "progress": progress
            }

            # Update tree status → processing
            _update_tree_status(file["path"], "processing")

            try:
                content = await github_raw(ri["owner"], ri["repo"], branch, file["path"])
                current_chunk += f"\n--- FILE: {file['path']} ---\n{content}\n"
                files_in_chunk.append(file["path"])
            except Exception:
                # Skip file không đọc được (binary, quá lớn, v.v.)
                pass

            # Update tree status → done
            _update_tree_status(file["path"], "done")

            # Flush chunk nếu đủ lớn hoặc là file cuối
            chunk_ready = len(current_chunk) >= 35_000 or i == total - 1
            if chunk_ready and current_chunk.strip() and files_in_chunk:
                DB["scan_status"]["text"] = (
                    f"AI đang vector hóa cụm {len(files_in_chunk)} file... "
                    f"({len(insights) + 1} insight)"
                )

                summary = await call_ai(
                    "INDEXER",
                    f"Tóm tắt logic code:\n{current_chunk}",
                    (
                        "Bạn là code indexer. Tóm tắt ngắn gọn chức năng, "
                        "patterns và dependencies của các file này. "
                        "KHÔNG bọc toàn bộ trong thẻ Markdown Code Block."
                    )
                )

                insight = {
                    "id": int(time.time() * 1000) + i,
                    "files": list(files_in_chunk),
                    "summary": summary
                }
                insights.append(insight)
                DB["knowledge_base"]["insights"] = list(insights)

                # Reset chunk
                current_chunk = ""
                files_in_chunk = []

        # --- Step 5: Project Wiki ---
        DB["scan_status"] = {
            "running": True,
            "text": "Đang viết Corporate Wiki (Project Summary)...",
            "progress": 90
        }

        all_insights_str = "\n\n".join(ins["summary"] for ins in insights)
        project_summary = await call_ai(
            "ARCHITECT",
            f"Dựa vào các mảnh ghép kỹ thuật này:\n{all_insights_str}",
            (
                f"Bạn là Chief Architect của dự án '{ri['repo']}'. "
                "Viết PROJECT SUMMARY chuẩn GitBook bao gồm các mục:\n"
                "# 📖 Introduction\n"
                "## 🧩 Core Components\n"
                "## ⚙️ Key Functions & APIs\n"
                "## 🗄️ Data Models\n"
                "## 🔐 Security Considerations\n"
                "## 🚀 Deployment & Infrastructure\n"
                "Trình bày Markdown sắc nét. "
                "BẮT BUỘC KHÔNG bọc toàn bộ kết quả trong thẻ ```markdown. "
                "TUYỆT ĐỐI KHÔNG bịa đặt thông tin."
            )
        )

        DB["knowledge_base"]["project_summary"] = project_summary
        DB["scan_status"] = {
            "running": False,
            "text": f"✅ Quét hoàn tất. Đã index {total} files, tạo {len(insights)} insights.",
            "progress": 100
        }

    except Exception as e:
        DB["scan_status"] = {
            "running": False,
            "text": f"❌ Lỗi Scan: {str(e)}",
            "progress": 0,
            "error": True
        }


def _update_tree_status(file_path: str, status: str):
    """Helper: cập nhật status của file trong knowledge_base.tree."""
    for f in DB["knowledge_base"]["tree"]:
        if f["path"] == file_path:
            f["status"] = status
            break

File: backend-ts/src/services/scan/scan.service.ts
/**
 * YUMMY Backend — Scan service.
 * Background task that scans a GitHub repo and indexes it into the knowledge base.
 *
 * Mirrors backend/services/scan_service.py.
 *
 * Flow:
 *   1. Fetch repo metadata (default branch)
 *   2. Fetch file tree, filter by extension and exclude node_modules / .git
 *   3. Read each file, group into ~35KB chunks
 *   4. Each chunk -> INDEXER agent summarizes
 *   5. All insights -> ARCHITECT agent writes Project Wiki
 *
 * Poll status via GET /kb/scan/status (handled by kb router).
 */
import { ALLOWED_EXTENSIONS, SCAN_CHUNK_BYTES } from '../../config/constants.js';
import type { Db } from '../../db/client.js';
import { kbRepo } from '../../db/repositories/kb.repo.js';
import { repoRepo } from '../../db/repositories/repo.repo.js';
import { scanStatusRepo } from '../../db/repositories/scan-status.repo.js';
import { callAI } from '../ai/dispatcher.js';
import { getRepoInfo, getRepoTree, githubRaw, type TreeEntry } from '../github/github.service.js';

const INDEXER_INSTRUCTION =
  'You are a code indexer. Briefly summarize the functionality, ' +
  'patterns and dependencies of these files. ' +
  'Do NOT wrap the entire output in a Markdown code block.';

function architectInstruction(repoName: string): string {
  return (
    `You are the Chief Architect of the '${repoName}' project. ` +
    'Write a PROJECT SUMMARY in GitBook style with the following sections:\n' +
    '# Introduction\n' +
    '## Core Components\n' +
    '## Key Functions & APIs\n' +
    '## Data Models\n' +
    '## Security Considerations\n' +
    '## Deployment & Infrastructure\n' +
    'Use clean Markdown formatting. ' +
    'Do NOT wrap the entire output in a ```markdown block. ' +
    'Do NOT fabricate any information.'
  );
}

function isAllowedFile(entry: TreeEntry): boolean {
  if (entry.type !== 'blob') return false;
  if (entry.path.includes('node_modules')) return false;
  if (entry.path.includes('.git')) return false;
  const ext = entry.path.slice(entry.path.lastIndexOf('.')).toLowerCase();
  return ALLOWED_EXTENSIONS.has(ext);
}

/**
 * Run a full repo scan in the background. Resolves when finished (or errored).
 * Status is persisted to scan_status; clients poll via GET /kb/scan/status.
 */
export async function runScan(db: Db): Promise<void> {
  await scanStatusRepo.set(db, {
    running: true,
    text: 'Connecting to GitHub API...',
    progress: 0,
    error: false,
  });
  await kbRepo.resetAll(db);

  try {
    const ri = await repoRepo.get(db);
    if (!ri) {
      await scanStatusRepo.set(db, {
        running: false,
        text: 'Scan error: repo info not configured.',
        progress: 0,
        error: true,
      });
      return;
    }
    const maxLimit = ri.maxScanLimit;

    // ── Step 1: repo metadata ──
    await scanStatusRepo.patch(db, { text: 'Reading repo info...' });
    const repoData = await getRepoInfo(db, ri.owner, ri.repo);
    const branch = repoData.default_branch;
    await repoRepo.setBranch(db, branch);

    // ── Step 2: file tree ──
    await scanStatusRepo.patch(db, { text: 'Fetching file list...' });
    const allFiles = await getRepoTree(db, ri.owner, ri.repo, branch);
    const validFiles = allFiles.filter(isAllowedFile).slice(0, maxLimit);

    if (validFiles.length === 0) {
      await scanStatusRepo.set(db, {
        running: false,
        text: 'No matching files found in the repo.',
        progress: 0,
        error: true,
      });
      return;
    }

    // Initialize tree with "pending" status
    await kbRepo.replaceTree(
      db,
      validFiles.map((f) => ({
        path: f.path,
        name: f.path.split('/').pop() ?? f.path,
        status: 'pending',
      })),
    );

    // ── Steps 3 & 4: read files in chunks, AI summarizes ──
    let currentChunk = '';
    let filesInChunk: string[] = [];
    let insightsCount = 0;
    const total = validFiles.length;

    for (const [i, file] of validFiles.entries()) {
      const progress = Math.round((i / total) * 80);
      await scanStatusRepo.set(db, {
        running: true,
        text: `Indexing [${file.path}] (${i + 1}/${total})`,
        progress,
        error: false,
      });

      await kbRepo.updateTreeStatus(db, file.path, 'processing');

      try {
        const content = await githubRaw(db, ri.owner, ri.repo, branch, file.path);
        currentChunk += `\n--- FILE: ${file.path} ---\n${content}\n`;
        filesInChunk.push(file.path);
      } catch {
        // Skip files that cannot be read (binary, too large, etc.)
      }

      await kbRepo.updateTreeStatus(db, file.path, 'done');

      // Flush chunk if large enough or last file
      const chunkReady = currentChunk.length >= SCAN_CHUNK_BYTES || i === total - 1;
      if (chunkReady && currentChunk.trim() && filesInChunk.length > 0) {
        await scanStatusRepo.patch(db, {
          text:
            `AI indexing chunk of ${filesInChunk.length} files... ` +
            `(${insightsCount + 1} insights so far)`,
        });

        const summary = await callAI(
          'INDEXER',
          `Summarize the code logic:\n${currentChunk}`,
          INDEXER_INSTRUCTION,
          undefined,
          db,
        );

        await kbRepo.addInsight(db, {
          id: Date.now() + i,
          files: [...filesInChunk],
          summary,
          createdAt: Date.now(),
        });
        insightsCount += 1;

        currentChunk = '';
        filesInChunk = [];
      }
    }

    // ── Step 5: project wiki ──
    await scanStatusRepo.set(db, {
      running: true,
      text: 'Writing Project Wiki (Project Summary)...',
      progress: 90,
      error: false,
    });

    const allInsightsStr = (await kbRepo.listInsights(db)).map((ins) => ins.summary).join('\n\n');

    const projectSummary = await callAI(
      'ARCHITECT',
      `Based on these technical summaries:\n${allInsightsStr}`,
      architectInstruction(ri.repo),
      undefined,
      db,
    );

    await kbRepo.setProjectSummary(db, projectSummary);
    await scanStatusRepo.set(db, {
      running: false,
      text: `Scan complete. Indexed ${total} files, generated ${insightsCount} insights.`,
      progress: 100,
      error: false,
    });
  } catch (e) {
    await scanStatusRepo.set(db, {
      running: false,
      text: `Scan error: ${(e as Error).message}`,
      progress: 0,
      error: true,
    });
  }
}
